Add seed trajectories to MotionPlanRequest

bmagyar opened the pull request:
Related to https://github.com/ros-planning/moveit/issues/804

Everything is up for debate.

My points so far:
* I used the `JointOrCartesianTrajectory` to avoid confusion regarding the contents of the message. Ideally `trajectory_msgs` could host this message.
* I chose `geometry_msgs/TransformStamped` to describe the points of a Cartesian trajectory to allow for any base frame to be used to define any point. This allows for multiple seeds even for different end effector(s) although leaves some questions regarding the properness of this arbitrary point. 
* Something like [this](https://github.com/bmagyar/industrial_moveit/blob/cartesian-seed/stomp_moveit/src/stomp_planner.cpp#L577) could be added to a conversion library to ease moving from-to the different spaces using the IK solver present.
* I added the `seed_trajectories` field to `MotionPlanRequest`.

Once we get to a good common ground I could try setting up a demo.
FYI: @rhaschke @davetcoleman @jrgnicho

bmagyar commented on msg/MotionPlanRequest.msg:
```diff
@@ -20,6 +20,9 @@ Constraints path_constraints
 # The constraints the resulting trajectory must satisfy
 TrajectoryConstraints trajectory_constraints
 
+# A set of trajectories to be used as seed trajectories for optimization-based planners
```
It definitely should!

bmagyar commented:
Thanks @davetcoleman for your feedback, you've brought up great point.

I'll keep adding commits to this PR to maintain a visible progression for anyone who wants to join in. 

Should we also add tolerances for `JointTrajectory` or only the Cartesian part?

bmagyar commented on msg/CartesianTrajectoryPoint.msg:
```diff
@@ -0,0 +1,14 @@
+geometry_msgs/Accel velocity_upper_tolerance
+geometry_msgs/Accel velocity_lower_tolerance
+
+geometry_msgs/Accel acceleration_upper_tolerance
```
Technically this is incorrect & inconsistent but didn't want to define Jerk.

bmagyar commented:
This version follows the pattern of `JointTrajectory` and `JointTrajectoryPoint` but I also included tolerances in the point.

davetcoleman commented on msg/CartesianTrajectory.msg:
```diff
@@ -0,0 +1,3 @@
+string eef_frame
```
could you add a comment at top explaining the intention of this file, and the others?

bmagyar replied:
Added everywhere

davetcoleman commented on msg/CartesianTrajectoryPoint.msg:
```diff
@@ -0,0 +1,14 @@
+geometry_msgs/Accel acceleration
+
+geometry_msgs/Twist pose_upper_tolerance
+geometry_msgs/Twist pose_lower_tolerance
```
I may be mathematically off here, but I think a tolerance for a pose is not a "twist"

Descartes simply defines tolerances as (x y z), (r p y)

https://github.com/ros-industrial-consortium/descartes/blob/kinetic-devel/descartes_trajectory/include/descartes_trajectory/cart_trajectory_pt.h

@mlautman thoughts?

bmagyar replied:
True, there is no time involved here.

davetcoleman commented on msg/CartesianTrajectoryPoint.msg:
```diff
@@ -0,0 +1,14 @@
+geometry_msgs/Twist pose_lower_tolerance
+
+geometry_msgs/Accel velocity_upper_tolerance
+geometry_msgs/Accel velocity_lower_tolerance
```
Twist?

bmagyar commented:
The thing I don't like about using `Pose` types for `pose_tolerance` is that it makes it hard to reason about a quaternion... A `Twist` would be more intuitive but then we are mixing apples and oranges...
What is you take on this?

bmagyar commented:
Ping

v4hn commented on msg/CartesianTrajectoryPoint.msg:
```diff
@@ -0,0 +1,14 @@
+geometry_msgs/Twist velocity_lower_tolerance
+
+geometry_msgs/Accel acceleration_upper_tolerance
+geometry_msgs/Accel acceleration_lower_tolerance
```
For all three of these groups I don't see how the lower-upper combination is useful really.
What is the "lower velocity tolerance" supposed to be? A lower limit?
Please feel free to disagree, but we are talking about derivatives of a 6 dimensional space here. Even the 6d pose space cannot be bound very well with only two bounds, not to talk about the derivatives. So you impose a one dimensional manifold with the current message which seems not very useful to me.

I would propose to reduce these fields to two numbers each for position tolerance (in m, m/s, m/s^2) and orientation tolerance (rad, rad/s, rad/s^2). This leaves us with a tolerance/uncertainty measure without imposing an artificial 1d interval.
If you want to represent your bounds then, you can still use three CartesianTrajectories, but personally I would go for Motion primitives instead.

bmagyar replied:
Indeed. The main reason for not using a pose or transform directly to describe the cartesian part was to address Dave's comment:
> This won't support tolerances, which is important for Descartes
I basically tried to mirror the Descartes limits as a first draft. Indeed, it is a lower limit. 

I slightly disagree with using a single limit as for planar plans for instance, one could allow higher tolerances on the Z axis, while having tight ones on X and Y. As the planners I'm working with doesn't really support tolerances, I don't have a strong position on this one.

davetcoleman replied:
@mlautman has worked with MoveIt! <> Descartes integration a lot recently, please weight in on the need for these

davetcoleman replied:
I agree we don't need lower and upper bounds for velocity and acceleration

I do think we need the ability to specify different tolerances between XYZ, RPY

mlautman replied:
Descartes uses pose tolerances but it doesn't have any velocity or acceleration upper or lower limits.

mlautman replied:
I think constraints are a little overkill. `CartesianTrajectoryPoint` is just a point in time and should not have a concept of tolerances. The MotionPlanRequest already has Constraints that can be used for tolerances

bmagyar commented on msg/GenericTrajectory.msg:
```diff
@@ -0,0 +1,3 @@
+Header header
+trajectory_msgs/JointTrajectory joint_trajectory
+geometry_msgs/TransformStamped[] cartesian_trajectory
```
addressed

bmagyar commented on msg/CartesianTrajectory.msg:
```diff
@@ -0,0 +1,3 @@
+string eef_frame
+
```
All true. I've renamed it to frame and added a small comment.

bmagyar commented on msg/CartesianTrajectoryPoint.msg:
```diff
@@ -0,0 +1,14 @@
+geometry_msgs/Accel acceleration
+
+geometry_msgs/Pose pose_upper_tolerance
+geometry_msgs/Pose pose_lower_tolerance
```
I'll resolve this as it will be part of the next comment.

bmagyar commented:
I shall build an example once this PR is ready to be merged to avoid having to update that one with every change here. I think the best candidate to add it to will be CHOMP and a unit test for that.

davetcoleman commented:
I like the idea of having an example implementation!

davetcoleman commented:
@jschleicher as we discussed in person, I think you'd have valuable input into the needs for this message and cartesian planning with moveit

davetcoleman commented on msg/CartesianTrajectory.msg:
```diff
@@ -0,0 +1,6 @@
+Header header
+
+# The name of the Cartesian frame the trajectory describes it's motion of.
+string frame
```
``Header`` already has a ``frame_id``, so we shouldn't need ``frame``.

bmagyar replied:
The header carries the reference frame id, this is the target. I think in the first version I called it `end effector frame` but decided to shorten it.

davetcoleman commented on msg/CartesianTrajectoryPoint.msg:
```diff
@@ -0,0 +1,14 @@
+geometry_msgs/Pose pose
```
please document what is optional, e.g.

> Which of these fields is necessary depends on the type of planner being used

bmagyar replied:
done

davetcoleman commented on msg/GenericTrajectory.msg:
```diff
@@ -0,0 +1,3 @@
+Header header
```
Document here that you'd use either joint_trajectory or cartesian_trajectory but not both

v4hn replied:
Why not? Whether or not that makes sense depends on the implementation using the message.

bmagyar replied:
I agree and @BryceStevenWilley a +1 to this too. 
What do you guys thing about merging this with the idea of potentially supporting several input trajectories of the same type?

In non-upstream work I handled supporting "optional" fields by adding a boolean flag to signify whether a field should be used or not but lately there's a new approach, using a vector. 

I propose to have 2 vectors instead of what there is currently. One for cartesian and one for joint space initial trajectories. There may or may not be one or more, it entirely depends on the one who receives it what to use and what to dispose of.

BryceStevenWilley replied:
:+1: for using multiple input trajectories of the same type. The vector approach is a little neater than a separate bool too.

bmagyar replied:
Vector approach implemented

davetcoleman commented on msg/CartesianTrajectory.msg:
```diff
@@ -0,0 +1,6 @@
+Header header
```
GenericTrajectory.msg already has a Header so I don't think we need it duplicated here

bmagyar replied:
I think `CartesianTrajectory` should be self-contained. At the same time, `trajectory_msgs/JointTrajectory` also has a header so there's an option of removing the header from `GenericTrajectory`. What do you think of that?

davetcoleman commented on msg/MotionPlanRequest.msg:
```diff
@@ -20,6 +20,9 @@ Constraints path_constraints
 TrajectoryConstraints trajectory_constraints
 
+# A set of trajectories to be used as reference or seed trajectories for planners
+GenericTrajectory[] reference_trajectories
```
Document how usage of this new ``reference_trajectories`` affects the previous approach to goal setting via ``Constraints[] goal_constraints``. Does it replace it? Supplement it?

bmagyar replied:
In my opinion they shouldn't replace them, I've amended the comment with this.

jschleicher commented:
Probably this heads into another direction than the original scope of this PR: I'd like to use `MotionPlanRequestMessage` for cartesian paths as well and encode the type of planning in the plannerId (and maybe PlanningOptions). That would supersede `computeCartesianPath` and put the functionality into a PlannerManager (what we did in pilz_industrial_motion). For Sequences of (Cartesian or joint-space or mixed) waypoints we use an array of MotionPlanRequests, see [pilz_industrial_motion package](https://github.com/PilzDE/pilz_industrial_motion/blob/kinetic-devel/pilz_msgs/msg/MotionSequenceItem.msg). 

For circular arcs we need an additional helper point (circle center), that we "hacky" encode in path_constraints, maybe that field could be generalized as well (see [python pilz_robot_programming](https://github.com/PilzDE/pilz_industrial_motion/blob/e601101d1a01c0214be1a36693e3f194b3be5acc/pilz_robot_programming/src/pilz_robot_programming/commands.py#L560) )

To sum it up we want to

- [x] plan to a (joint or cartesian) goal and have a cartesian linear path (nothing new)
- [ ] plan through multiple such points with different planner (sub-)types ("Sequence")
- [ ] blend in between the subparts of a sequence
- [ ] add constraints that the trajectory doesn't need to plan through (but keep a specified distance in the case of a circle)

bmagyar commented:
Hi @jschleicher , I agree with you. The scope of this PR is to pass in an initial/seed trajectory to start an optimization-based planner from. The things you suggest however are somewhat related and will be partly enabled by this PR however I think we should follow it up with an extension to `PlanningOptions` which could determine the interpretation of the cartesian trajectory that was passed in such that:
* plan through points (new flag)
* determine how to blend between these point-to-point plans (new flag)
* add no-go areas (could be a new set of constraints that are enabled when following points, or when a new flag is enabled)

davetcoleman commented:
@bmagyar we (@ommmid ) are working on TrajOpt and need a way to pass in a seed joint-space trajectory. This PR is perfect for that, do you think you could address the remaining feedback so we can merge this? A lot of it is simply documentation requests

davetcoleman commented:
I've just targeted this to the master branch.

bmagyar commented:
Thanks @davetcoleman , I'll move forward with addressing the comments but unfortunately due to impeding deadlines regarding my thesis I won't be able to supply an example. It sound like the TrajOpt effort would be perfectly fit for this purpose, right?

ommmid commented:
@bmagyar yes, that is exactly what we need in trajopt.

davetcoleman commented:
friendly ping?

v4hn commented on msg/GenericTrajectory.msg:
```diff
@@ -0,0 +1,3 @@
+Header header
+trajectory_msgs/JointTrajectory joint_trajectory
```
why allow for multiple Cartesian trajectories, but only one joint trajectory?

bmagyar replied:
See above
https://github.com/ros-planning/moveit_msgs/pull/46/files#r328268508

BryceStevenWilley replied:
In addition, this allows you to have some separate semantic distinctions between paths (i.e., all of trajectories in the Generic trajectory go to the same goal, and can be used in combination during whatever optimization step happens), whereas each Generic trajectory is a separate target.

bmagyar replied:
And that is what I'm a little worried about. Maybe this is something we should mention, as having a vector of 2 vectors can quickly become unclear.

bmagyar replied:
I added the vectorization to GenericTrajectory and a comment on semantic equivalence

mlautman commented on msg/CartesianTrajectory.msg:
```diff
@@ -0,0 +1,6 @@
+Header header
+
+# The name of the Cartesian frame the trajectory describes it's motion of.
```
```suggestion
# The name of the target Cartesian frame being tracked with respect to the base frame provided in header.frame_id
```

bmagyar commented:
ok, I am finally out of the grasp of PhD work, let's get this baby delivered!
I'll resolve what was done, whoever wishes to review, feel free to re-open if I closed something by mistake.

@mlautman @davetcoleman can we aim to merge this by Friday?

mlautman commented on CMakeLists.txt:
```diff
@@ -18,13 +18,16 @@ AllowedCollisionEntry.msg
 AllowedCollisionMatrix.msg
 AttachedCollisionObject.msg
 BoundingVolume.msg
+CartesianTrajectory.msg
```
```suggestion
CartesianPoint.msg
CartesianTrajectory.msg
```

bmagyar replied:
done!

bmagyar replied:
done

bmagyar commented on CMakeLists.txt:
```diff
@@ -18,13 +18,16 @@ AllowedCollisionEntry.msg
 AttachedCollisionObject.msg
 BoundingVolume.msg
+CartesianTrajectory.msg
+CartesianTrajectoryPoint.msg
```
done

bmagyar commented:
All requested changes implemented, commits squashed :+1:
